x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/index.tsx

```diff
@@ -392,21 +392,23 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
   }, [onSubmit]);
 
   const getData = useCallback(async () => {
-    const result = await submit();
-    result.data = {
-      ...result.data,
-      eqlOptions: optionsSelected,
+    // validate doesn't return actual state of form
+    // more details here: https://github.com/elastic/kibana/issues/144322#issuecomment-1321838136
+    // wrapping in setTimeout is a workaround until solution within forms-lib can be found
+    const isValid = await new Promise<boolean>((resolve) => {
+      setTimeout(async () => {
```

Comment:
The fix looks dirty but fine if it's temporary.

I just wanted to share my thoughts. Usually a necessity of `setTimeout()` usage says about design flaws and over-complication. In this case it looks possible to get rid of the problem by rethinking the approach to the data handling. For example I can see `formHooks` is used as a global variable and the value is reused the level up.
[The cause of the problem](https://github.com/elastic/kibana/issues/144322#issuecomment-1321838136) is double validation which actually shouldn't happen.

I'm curious if an approach to redesign the components to handle validation in another way was considered as well.

Reply:
> The fix looks dirty but fine if it's temporary.
I would recommend to read [this section](https://www.elastic.co/guide/en/kibana/master/pr-review.html#_inappropriate_review_feedback) before writing comments like this.

> I just wanted to share my thoughts. Usually a necessity of setTimeout() usage says about design flaws and over-complication. In this case it looks possible to get rid of the problem by rethinking the approach to the data handling. For example I can see formHooks is used as a global variable and the value is reused the level up.
https://github.com/elastic/kibana/issues/144322#issuecomment-1321838136 is double validation which actually shouldn't happen.
> I just wanted to share my thoughts. Usually a necessity of setTimeout() usage says about design flaws and over-complication. In this case it looks possible to get rid of the problem by rethinking the approach to the data handling. For example I can see formHooks is used as a global variable and the value is reused the level up.
https://github.com/elastic/kibana/issues/144322#issuecomment-1321838136 is double validation which actually shouldn't happen.

What do you mean by 'double validation'?
I would expect once data in form is changed and available in onChange callback, it can be checked whether it's valid or not. But if you look into standalone [example](https://github.com/elastic/kibana/issues/145846), it's not a case: data is changed, but its validity can not be established.
I would like to see forms-lib assessment of library behaviour before committing to any refactoring.

Reply:
Thank you for the explanation and pointing to the guidelines.

I didn't mean to offend but rather say that `setTimeout` usually is a red flag which signals there is something wrong.

You are right the problem in the the forms-lib. Though it's strange the popular scenario of accessing the validity state in the `onChange` callback wasn't covered already.

Reply:
> rather say that setTimeout usually is a red flag which signals there is something wrong.

yes,  I also commented that it's a workaround and put together ticket and links to it

>     // wrapping in setTimeout is a workaround until solution within forms-lib can be found


> You are right, the problem is in the forms-lib. Though it's strange the popular scenario of accessing the validity state in the onChange callback wasn't covered already.
In this [standalone example ](https://github.com/elastic/kibana/issues/145846), it can be seen that in onChange, accessed validity is not correct. Very simple case, but the result is rather unexpected. Let's see what forms-lib will say on that.
I would expect if I get any data from callback: to either access validation status straight away or through running `validate` method. We did the latter in form(though through calling a `submit` before this change) and relied on its results. Which, as it turns out, are not correct in some cases.